### PR TITLE
feat(MCS Library): add other data channel types

### PR DESCRIPTION
**Pull request by evenchange4**
~- [ ] Is it okay to separate Class for different DataChannel? one per file~
- [x] How to test?

## SOP
1. Copy paste
2. rename
3. type
4. default value
5. parse
6. toString

## DataChannel List

- [x] Category
- [x] Integer
- [x] Float
- [x] Hex
- [x] String
- [x] GPS
- [x] GPIO
- [x] PWM
- [x] Analog
- [x] Gamepad

## Test

- [x] compile
- [x] end-to-end

### CODE REVIEW

**Comment by pablosun** [comment]
@SmallpTsaiMTK Please help to comment on Michael's questions, thanks.

File: middleware/third_party/arduino/hardware/arduino/mt7697/libraries/MCS/src/MCS.cpp [hunk 1]
```diff
@@ -483,4 +483,92 @@ bool MCSDisplayOnOff::value(void)
...
+}
+
+
+/* ----------------------------------------------------------------------------
+class MCSControllerInteger
+---------------------------------------------------------------------------- */
+
+MCSControllerInteger::MCSControllerInteger(const String& channel_id):
+MCSDataChannel(channel_id),
+mValue(0) // TODO: Intial value ?? N/A ??
```
Comment (evenchange4) [inline middleware/third_party/arduino/hardware/arduino/mt7697/libraries/MCS/src/MCS.cpp]: I am not sure about setting default value. Zero might not be the lower bound.

Comment (SmallpTsai) [inline middleware/third_party/arduino/hardware/arduino/mt7697/libraries/MCS/src/MCS.cpp]: mValue is meaning less, before mInited set to true.

Comment (pablosun) [inline middleware/third_party/arduino/hardware/arduino/mt7697/libraries/MCS/src/MCS.cpp]: Still 0 is generally a good idea for default values. A common C idiom is that numerical members are always initialized to 0 as default, and is supported by [some language syntax](http://en.cppreference.com/w/cpp/language/zero_initialization).

File: middleware/third_party/arduino/hardware/arduino/mt7697/libraries/MCS/src/MCS.cpp [hunk 2]
```diff
@@ -492,7 +492,7 @@ class MCSControllerCategory
 
 MCSControllerCategory::MCSControllerCategory(const String& channel_id):
 MCSDataChannel(channel_id),
-mValue('')
+mValue('0')
```
Comment (pablosun) [inline middleware/third_party/arduino/hardware/arduino/mt7697/libraries/MCS/src/MCS.cpp]: ## Suggest removing `mValue` from the initialization list (use its default constructor)

因為 `mValue` 的 Type 是 `String` 這個 Class，它已經有提供 default constructor，所以建議不用再特別 initialize 成 `'0'` 這個字串。建議這邊可以直接把 `mValue` 從 initialization list移除。

附註一：`String` 這個 class 有特別設計過，如果還沒有被賦值過，他會 evaluate 成 `false`。

附註二：C的字串是 `char *`，一般也是初始化為 NULL（數值為`0`），而不是 `'0'`這個字元 (`'0'`事實上是一個 ASCII value，數值是 `48` ）

**Comment by evenchange4** [comment]
## Update

Ready to review

- [ ] Hex: This should be of integral types, rather than a string

=> https://github.com/MediaTek-Labs/Arduino-Add-On-for-LinkIt-SDK/pull/6/commits/4134d00fae6ce5136eb522f6079f8a03dafc280d

- [ ] GPS: It should be have interfaces like getLat and getLong, or the getValue method should return a Location object that supports getLat and getLong

=> https://github.com/MediaTek-Labs/Arduino-Add-On-for-LinkIt-SDK/pull/6/commits/b4302f0a197b6f46aa1efb8b67d7f3e4377ccb83

- [ ] PWM: It should have methods like getDutyCycle and getFrequency instead of a single getValue function

=> https://github.com/MediaTek-Labs/Arduino-Add-On-for-LinkIt-SDK/pull/6/commits/d65f09135bd986ee03d59563059008d6c9e0b185

- [ ] GamePad: the get/set methods should be an enum or some object that can represent button state of the game pad, instead of a free-form string. Current interface allows user to input arbitrary strings, which can easily violate the design of the GamePad data channel.

=> https://github.com/MediaTek-Labs/Arduino-Add-On-for-LinkIt-SDK/pull/6/commits/ca97cd1a0d2cb329c3827a2ffbd53691e3d2eb48

File: middleware/third_party/arduino/hardware/arduino/mt7697/libraries/MCS/src/MCS.cpp [hunk 3]
```diff
@@ -787,7 +787,7 @@ void MCSControllerHex::_dispatch(const String& params)
 bool MCSControllerHex::_update(const String& params)
 {
     String b = params;
-    String v = b;
+    long v = strtol(&b[0], NULL, 16);
```
Comment (SmallpTsaiMTK) [inline middleware/third_party/arduino/hardware/arduino/mt7697/libraries/MCS/src/MCS.cpp]: you can use String::c_str() to get char* pointer
WString.h
const char* c_str() const { return buffer; }

**Comment by evenchange4** [comment]
## Gamepad discussion

1. Should we support for Gamepad?

    The **Gamepad** data channel could only work with TCP connection.
For `long-polling` mode (restful-api), we will lose some commands from the MCS console.

2. Gamepad values are dynamic.

     If user press up button, the value will be `a|1`. Is there any suggestion for handling dynamic state of Gamepad values?



cc @SmallpTsai @iamblue

File: middleware/third_party/arduino/hardware/arduino/mt7697/libraries/MCS/src/MCS.cpp (hunk 3 above)
Comment (evenchange4) [inline middleware/third_party/arduino/hardware/arduino/mt7697/libraries/MCS/src/MCS.cpp]: done

**Comment by pablosun** [comment]
@evenchange4 

1. I think this should be considered as a **bug** for MCS? (we will lose some commands from the MCS console)

2. 
## Assigning key values
The best answer is that MCS provides the "meta-data" for the key-value setting.

If this is not possible, I would suggest that the constructor/init method for `Gamepad` class requires user to assign key values so that the class could parse the value for them. The user is responsible for making sure the key values passed to constructors are exactly the same as the MCS setting, e.g.
```cpp
MCSControllerChannelGamepad(const String& channel_id, char UP, char DOWN, char LEFT, char RIGHT, char BUTTON_A, char BUTTON_B);
```

## For the return values
The key combination is a set of finite element that can be combined together. This can be represented by **Bit Field** in C/C++, e.g. each direction/button is represented by a bit in the field, e.g.
```c
UP = 01b
LEFT = 10b
```
Then user can use bit-wise AND operation to check if a certain key is pressed:
```c
if( UP | keyInfo )
{
    // Up is pressed
}
```

**Comment by pablosun** [comment]
Also, please fix the build error reported by Travis.
(All non-void functions must always **return** some value).

**Comment by evenchange4** [comment]
done

- [ ] remove Gamepad
- [ ] All non-void functions must always return some value

cc @pablosun @SmallpTsaiMTK

**Pull request merged**